src/core/midiPorts.cpp

```diff
@@ -0,0 +1,416 @@
+
+int getInDeviceIndex(std::string port){
+
+	std::vector<std::string> devices = getInDevices();
+
+	auto it = find(devices.begin(), devices.end(), port);
+	if (it !=devices.end()) {
+		return distance(devices.begin(), it);
+	}
+	return -1;
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+int openOutPort(std::string port){
+
+	// First let's see if port name is even valid
+	int index = getOutDeviceIndex(port);
+	if (index < 0) {
+		// No such port is available, exiting
+		return -1;
+	}
+
+	// Let's check if this port is in the map, possibly already open //
+	if (outPorts_.count(port) > 0) {
+
+		// This port already exists in the map
+		if (outPorts_[port]->isPortOpen()) {
+		
+			// And this port is indeed open, nothing to do.
+			return 2;
+		}
+		// However if it is in the map but not open, something is wrong
+		// Oh well, we'll se what happens next.
+	}
+
+	// No entry in map, so we shall create one
+	else {
+		try {
+			outPorts_[port] = new RtMidiOut((RtMidi::Api) api_,
+						"Giada Output to " + port);
+		}
+		catch (RtMidiError& error) {
+			u::log::print("[MP::openOutPort] RtMidiError: %s\n",
+						error.getMessage().c_str());
+			return -2;
+		}
+	}
+	
+	// Let's open this port
+	try {
+		outPorts_[port]->openPort(index, port);
+		u::log::print("[MP::openOutPort] MIDI port \"%s\" open\n", port);
+		return 1;
+	}
+	catch (RtMidiError& error) {
+		u::log::print("[MP::openOutPort] unable to open port %s: %s\n",
+					port, error.getMessage().c_str());
+		return -3;
+	}
+
+}
+
+/* -------------------------------------------------------------------------- */
+
+int openInPort(std::string port){
+
+	// First let's see if port name is even valid
+	int index = getInDeviceIndex(port);
+	if (index < 0) {
+		// No such port is available, exiting
+		return -1;
+	}
+
+	// Let's check if this port is in the map, possibly already open //
+	if (inPorts_.count(port) > 0) {
+
+		// This port already exists in the map
+		if (inPorts_[port]->isPortOpen()) {
+		
+			// And this port is indeed open, nothing to do.
+			return 2;
+		}
+		// However if it is in the map but not open, something is wrong
+		// Oh well, we'll se what happens next.
+	}
+
+	// No entry in map, so we shall create one
+	else {
+		try {
+			inPorts_[port] = new RtMidiIn((RtMidi::Api) api_,
+						"Giada Input to " + port);
+		}
+		catch (RtMidiError& error) {
+			u::log::print("[MP::openInPort] RtMidiError: %s\n",
+						error.getMessage().c_str());
+			return -2;
+		}
+	}
+	
+	// Let's open this port
+	try {
+		// Create a persistent string with a port name, so callback
+		// function knows which port did the data come from. 
+		// TODO: How do we dispose of this thing when the port is closed?
+		// A unique_ptr would be nice but I think it's not applicable here.
+		std::string* p = new std::string(port);
+
+		inPorts_[port]->setCallback(&callback_, p);
```

Comment:
off the top of my head: passing a lambda as callback might help to avoid the string allocation. That would require to change the setCallback signature of course.

Reply:
You mean lambda function that would always return the same string value? Okay, I'll look into that.

What do you mean by "changing setCallback signature"?

Reply:
> What do you mean by "changing setCallback signature"?

My bad, I didn't notice that `inPorts_[port]` is RtMidi stuff, so the signature can't be changed (we have no control over it). I'm still 50% sure you can pass a lambda in there. I'll try to dig deeper.

Reply:
Okay, I'll let you think about this for some time before I try to solve it myself. :)

Reply:
OK, my theory was wrong :). Question: what if you pass the port *number* instead and reconstruct the name in the callback function?

Reply:
I decided to use port names in string format because numbering is very unreliable - list of ports teturned by RtMidi changes as you run and close other midi-capable applications. I experienced many problems because of that, and this is why I implemented [this](https://github.com/monocasual/giada/pull/398/commits/32051640cca1de45d0666afb6ac652752324bdbd) right from the start. :) 
Numbering items in the ports map is no good either, as map will change during runtime as well.

Well, if we don't find a cleaner solution, I'll probably introduce yet another map just for port names, so callbacks can safely operate with numerical data disguised as (void *). Just wanted to make it cleaner somehow. :)

Reply:
And what would my_real_callback look like in this case? what is that +[&] thing?

Reply:
The `+` is a unary conversion operator for a lambda to just a function pointer, if you need it.
Kind of a hammer to force some modern code into old C API... :-) https://stackoverflow.com/questions/17822131/resolving-ambiguous-overload-on-function-pointer-and-stdfunction-for-a-lambda
Just try without it at the first place.

Reply:
See my other comment where c`allback_` is declared